File: javascript/Phylo/Parsers/Nexml.js
(function(){
function parse(args) {
	var xml = args["string"];
	var xmlDoc;
	var result = new Array();
	try { //Internet Explorer
  		xmlDoc = new ActiveXObject("Microsoft.XMLDOM");
  		xmlDoc.async = "false";
  		xmlDoc.loadXML(text);
  	}
	catch(e) {
  		try { //Firefox, Mozilla, Opera, etc.
  			var parser = new DOMParser();
  			xmlDoc = parser.parseFromString(xml,"text/xml");
  		}
  		catch(e) {
  			alert(e.message);
  			return result;
  		}
	}
	var taxa_blocks = process_otus(xmlDoc.getElementsByTagName("otus"));
	var tree_blocks = process_trees(xmlDoc.getElementsByTagName("trees"));
	var char_blocks = process_characters(xmlDoc.getElementsByTagName("characters"));
	resolve_tree_taxa(taxa_blocks,tree_blocks);
	resolve_matrix_taxa(taxa_blocks,char_blocks);
	for ( var key in taxa_blocks ) {
		result.push(taxa_blocks[key]);
	}
	for ( var i in char_blocks ) {
		result.push(char_blocks[i]);
	}
	for ( var i in tree_blocks ) {
		result.push(tree_blocks[i]);
	}
	return result;
}
Phylo.Parsers.Nexml.parse = parse;

function resolve_matrix_taxa(taxa_blocks,char_blocks) {
	for ( var i in char_blocks ) {
		var taxa_id = char_blocks[i].get_generic('otus');
		var taxa = taxa_blocks[taxa_id];
		var taxon = {};
		taxa.visit(
			function(t) {
				var taxon_id = t.get_xml_id();
				taxon[taxon_id] = t;
			}
		);
		var rows = char_blocks[i].get_entities();
		for ( var j in rows ) {
			var taxon_id_ref = rows[j].get_generic('otu');
			rows[j].set_taxon(taxon[taxon_id_ref]);
		}
		char_blocks[i].set_taxa(taxa);
	}
}

function process_characters(char_elts) {
	var char_blocks = [];
	if(char_elts.length==0) return char_blocks;
	for ( var i = 0; i < char_elts.length; i++ ) {
		var char_obj = obj_from_elt(char_elts[i]);
		var type = char_elts[i].attributes.getNamedItem('xsi:type').value;
		type = type.match(/^nex:([A-Z][a-z]+)/)[1];
		char_obj.set_type(type);
		var otus = char_elts[i].attributes.getNamedItem('otus').value;
		char_obj.set_generic({ 'otus' : otus });
		var format_elt = char_elts[i].getElementsByTagName('format')[0];
		/*
		 * state_set = {
		 *     'states1' : { // state set ID
		 *         'symbols' : {
		 *             's1' : 1, // i.e. state ID to state symbol mapping
		 *             's2' : 2
		 *         }
		 *         lookup : {
		 *             1 : [ 1 ], // ambiguity mapping
		 *             2 : [ 2 ],
		 *         }
		 *     },
		 *     'states2' : {
		 *         // etc.
		 *     }
		 * }
		 */
		var state_set = {};
		var state_set_of_char = {}; // char id to state set mapping
		var col_indices = {};
		if ( format_elt ) {			
			var states_elts = format_elt.getElementsByTagName('states');
			for ( var j = 0; j < states_elts.length; j++ ) {
				var id = states_elts[j].attributes.getNamedItem('id').value;
				var lookup = {};
				var symbol_of_id = {};
				var state_elts = states_elts[j].getElementsByTagName('state');
				var pss = states_elts[j].getElementsByTagName('polymorphic_state_set');
				var uss = states_elts[j].getElementsByTagName('uncertain_state_set');
				resolve_mapping(state_elts,symbol_of_id,lookup);
				resolve_mapping(pss,symbol_of_id,lookup);
				resolve_mapping(uss,symbol_of_id,lookup);
				state_set[id] = { 'lookup' : lookup, 'symbols' : symbol_of_id };
				char_obj.set_lookup(lookup);
			}
			var col_elts = format_elt.getElementsByTagName('char');
			for ( var j = 0; j < col_elts.length; j++ ) {
				var id = col_elts[j].attributes.getNamedItem('id').value;
				var states = col_elts[j].attributes.getNamedItem('states').value;
				state_set_of_char[id] = states;
				col_indices[id] = j;
			}
		}
		var matrix_elt = char_elts[i].getElementsByTagName('matrix')[0];
		process_matrix(char_obj,matrix_elt,state_set_of_char,state_set,col_indices);
		char_blocks.push(char_obj);
	}	
	return char_blocks;
}

function process_matrix(char_obj,matrix_elt,state_set_for_char,state_set,col_indices) {
	var rows = matrix_elt.getElementsByTagName('row');
	for ( var i = 0; i < rows.length; i++ ) {
		var row_obj = obj_from_elt(rows[i]);
		var otu = rows[i].attributes.getNamedItem('otu').value;
		row_obj.set_type_object(char_obj.get_type_object());		
		row_obj.set_generic({'otu':otu});
		var seq_elt = rows[i].getElementsByTagName('seq')[0];
		if ( seq_elt ) {
			row_obj.set_char([seq_elt.textContent]);
		}
		else {
			var cell_elts = rows[i].getElementsByTagName('cell');
			var characters = [];
			for ( var j = 0; j < cell_elts.length; j++ ) {
				var char_id = cell_elts[j].attributes.getNamedItem('char').value;
				var col_index;
				if ( col_indices[char_id] != null ) {
					col_index = col_indices[char_id]; 
				}
				else {
					col_index = char_id;
				}
				var state_id = cell_elts[j].attributes.getNamedItem('state').value;
				var state_set_id = state_set_for_char[char_id];
				var state;
				if ( state_set_id != null ) {
					state = state_set[state_set_id]['symbols'][state_id];
				}
				else {
					state = state_id;
				}
				characters[col_index] = state;
			}
			var missing = row_obj.get_missing();
			for ( var j in characters ) {
				if ( characters[j] == null ) {
					characters[j] = missing;
				}
			}
			row_obj.set_char(characters);
		}
	}	
}

function resolve_mapping(state_elts,symbol_of_id,lookup) {
	for ( var i = 0; i < state_elts.length; i++ ) {
		var state_elt = state_elts[i];
		var symbol = state_elt.attributes.getNamedItem('symbol').value;
		var id = state_elt.attributes.getNamedItem('id').value;
		symbol_of_id[id] = symbol;
		var member_elts = state_elt.getElementsByTagName('member');
		if ( member_elts.length ) {
			var symbols = [];
			for ( var i in member_elts ) {
				var member_id = member_elts[i].attributes.getNamedItem('state').value;
				symbols.push(symbol_of_id[member_id]);
			}
			lookup[symbol] = symbols;
		}
		else {
			lookup[symbol] = [symbol];
		}
	}
}

function resolve_tree_taxa(taxa_blocks,tree_blocks) {
	for ( var i = 0; i < tree_blocks.length; i++ ) {
		var otus_id = tree_blocks[i].get_generic('otus');
		var taxa_block = taxa_blocks[otus_id];
		tree_blocks[i].set_taxa(taxa_block);
		var taxa = taxa_block.get_entities();
		var taxa_by_id = {};
		for ( var j = 0; j < taxa.length; j++ ) {
			taxa_by_id[taxa[j].get_xml_id()] = taxa[j];
		}
		var trees = tree_blocks[i].get_entities();
		for ( var j = 0; j < trees.length; j++ ) {
			var nodes = trees[j].get_entities();
			for ( var k = 0; k < nodes.length; k++ ) {
				var id = nodes[k].get_generic('otu');
				if ( id != null ) {
					nodes[k].set_taxon(taxa_by_id[id]);
				}
			}
		}
	}
}
Phylo.Parsers.Nexml.resolve_tree_taxa = resolve_tree_taxa;

function process_trees(trees_elt) {
	var result = new Array();
	if (!trees_elt) return result;
	for ( var i = 0; i < trees_elt.length; i++ ) {
		var forest_obj = obj_from_elt(trees_elt[i]);
		var otus_id = trees_elt[i].attributes.getNamedItem("otus").value;
		forest_obj.set_generic({'otus':otus_id});
		var tree_elt = trees_elt[i].getElementsByTagName("tree");
		for ( var j = 0; j < tree_elt.length; j++ ) {
			forest_obj.insert(process_tree(tree_elt[j]));
		}
		result.push(forest_obj);
	}
	return result;
}
Phylo.Parsers.Nexml.process_trees = process_trees;

function process_tree(tree_elt) {
	var tree_obj = obj_from_elt(tree_elt);
	var source_of  = {};
	var target_of  = {};
	var length_of  = {};
	var otu_of     = {};
	var node_by_id = {};
	var node_elt = tree_elt.getElementsByTagName("node");
	for ( var i = 0; i < node_elt.length; i++ ) {
		var node_obj = obj_from_elt(node_elt[i]);
		var id = node_obj.get_xml_id();
		tree_obj.insert(node_obj);
		node_by_id[id] = node_obj;
		var otuAttr = node_elt[i].attributes.getNamedItem("otu");
		if (otuAttr != null) {
			var otu = otuAttr.value;
			otu_of[id] = otu;
			node_obj.set_generic({ 'otu' : otu });
		}
	}
	var edge_elt = tree_elt.getElementsByTagName("edge");
	for ( var i = 0; i < edge_elt.length; i++ ) {
		var attrs  = edge_elt[i].attributes;
		var target = attrs.getNamedItem("target").value;
		var source = attrs.getNamedItem("source").value;		
		source_of[target] = source;
		target_of[source] = target;
		if ( attrs.getNamedItem("length") != null ) {
			length_of[target] = attrs.getNamedItem("length").value;
		}
	}
	var entities = tree_obj.get_entities();
	for ( var i = 0; i < entities.length; i++ ) {
		var id = entities[i].get_xml_id();
		if ( source_of[id] != null ) {
			var parent_node = node_by_id[source_of[id]];
			entities[i].set_parent(parent_node);
			parent_node.set_child(entities[i]);
		}
		if ( length_of[id] != null ) {
			entities[i].set_branch_length(length_of[id]);
		}
		if ( otu_of[id] != null ) {
			entities[i].set_generic({ 'otu' : otu_of[id] });
		}
	}
	var root_edge = tree_elt.getElementsByTagName("rootedge")[0];
	if ( root_edge != null ) {
		var target = root_edge.attributes.getNamedItem("target").value;
		var bl = root_edge.attributes.getNamedItem("length").value;
		node_by_id[target].set_branch_length(bl);
	}	
	return tree_obj;
}
Phylo.Parsers.Nexml.process_tree = process_tree;

function process_otus (otus) {
	var taxa_blocks = {};
	for ( var i = 0; i < otus.length; i++ ) {
		var taxa = obj_from_elt(otus[i]);
		var otu = otus[i].getElementsByTagName("otu");
		for ( var j = 0; j < otu.length; j++ ) {
			var taxon = obj_from_elt(otu[j]);
			taxa.insert(taxon);
		}
		taxa_blocks[taxa.get_xml_id()] = taxa;
	}
	return taxa_blocks;	
}
Phylo.Parsers.Nexml.process_otus = process_otus;

function obj_from_elt (elt) {
	var attrs = elt.attributes;
	var args = {};
	for ( var j = 0; j < attrs.length; j++ ) {
		if ( attrs[j].name == 'label' && attrs[j].value != '' ) {
			args["name"] = attrs[j].value;
		}
		else if ( attrs[j].name == 'id' && attrs[j].value != '' ) {
			args["xml_id"] = attrs[j].value;
		}
	}
	var children = elt.childNodes;
	for ( var i = 0; i < children.length; i++ ) {
		if ( children[i].nodeType == 1 && children[i].tagName == 'dict' ) {
			args["generic"] = { 'dict' : parse_dict(children[i]) };
			break;
		}
	}
	var tag_name = elt.nodeName.toLowerCase();
	switch(tag_name) {
		case 'otus'       : return new Phylo.Taxa(args);
		case 'otu'        : return new Phylo.Taxa.Taxon(args);
		case 'trees'      : return new Phylo.Forest(args);
		case 'tree'       : return new Phylo.Forest.Tree(args);
		case 'node'       : return new Phylo.Forest.Node(args);
		case 'characters' : return new Phylo.Matrices.Matrix(args);
		case 'row'        : return new Phylo.Matrices.Datum(args);
		default : throw new Phylo.Util.Exceptions.API("Can't create object from element " + tag_name);
	}
}
Phylo.Parsers.Nexml.obj_from_elt = obj_from_elt;

function parse_dict (elt) {
	var result = {};
	var children = elt.childNodes;
	for ( var i = 0; i < children.length; i++ ) {		
		if ( children[i].nodeType == 1 && children[i].tagName == 'key' ) {
			var key = children[i].textContent;
			var value_elt;
			for ( var j = i+1; j < children.length; j++ ) {
				if ( children[j].nodeType == 1 ) {
					value_elt = children[j];
					break;
				} 
			}
			var regex = /vector/;
			var vector = value_elt.tagName.match(regex);
			if ( value_elt.tagName == 'dict' ) {
				result[key] = parse_dict(value_elt);
			} 
			else if ( vector != null ) {
				var value = new Array();
				value.push(value_elt.tagName);
				var value_array = value_elt.textContent.split(/\s+/);
				for ( var i = 0; i < value_array.length; i++ ) {
					value.push(value_array[i]);
				}
				result[key] = value_array;
			}
			else if ( value_elt.tagName == 'any' ) {
				result[key] = [ 'any', value_elt ];
			}
			else {
				result[key] = [ value_elt.tagName, value_elt.textContent ];
			}
		}
	}
	return result;	
}
Phylo.Parsers.Nexml.parse_dict = parse_dict;
})()

File: nexml/java/src/org/biophylo/Parsers/Nexml.java
package org.biophylo.Parsers;
import javax.xml.parsers.*;
import java.io.*;
import org.w3c.dom.*;
import java.util.*;
import org.biophylo.Util.*;
import org.biophylo.Taxa.*;
import org.biophylo.Forest.*;
import org.biophylo.Util.Exceptions.*;
import org.biophylo.*;
import org.biophylo.Matrices.*;
import org.biophylo.Matrices.Datatype.*;
import org.biophylo.Mediators.*;
public class Nexml implements Parsers {
	private static HashMap factory = null;
	private static ObjectMediator om = ObjectMediator.getInstance();
	private static Logger logger = Logger.getInstance();
	
	public Nexml () {
		if ( factory == null ) {
			factory = new HashMap();
			factory.put("otus", org.biophylo.Taxa.Taxa.class);
			factory.put("otu", org.biophylo.Taxa.Taxon.class);
			factory.put("characters", org.biophylo.Matrices.Matrix.class);
			factory.put("row", org.biophylo.Matrices.Datum.class);
			factory.put("trees", org.biophylo.Forest.Forest.class);
			factory.put("tree", org.biophylo.Forest.Tree.class);
			factory.put("node", org.biophylo.Forest.Node.class);
		}
	}
	
	public Object[] parse(InputStream data) {
		DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
		DocumentBuilder db = null;
		Document xmlDoc = null;
		Object[] theBlocks = null;
		try {
			db = dbf.newDocumentBuilder();
		} catch (ParserConfigurationException e) {
			e.printStackTrace();
		}
		try {			
			xmlDoc = db.parse(data);
		} catch (Exception e) {
			e.printStackTrace();
		}
		try { 
			HashMap taxaBlocks = processOtus(xmlDoc.getElementsByTagName("otus"));
			Vector treeBlocks = processTrees(xmlDoc.getElementsByTagName("trees"));
			Vector charBlocks = processCharacters(xmlDoc.getElementsByTagName("characters"));
			resolveTreeTaxa(taxaBlocks,treeBlocks);
			resolveMatrixTaxa(taxaBlocks,charBlocks);
			int numBlocks = taxaBlocks.size();
			numBlocks += treeBlocks.size();
			numBlocks += charBlocks.size();
			Object[] blocks = new Object[numBlocks];
			Object[] taxa = taxaBlocks.values().toArray();
			for ( int i = 0; i < taxa.length; i++ ) {
				blocks[i] = taxa[i];
			}
			for ( int i = 0; i < treeBlocks.size(); i++ ) {
				blocks[i+taxa.length] = treeBlocks.get(i);
			}
			for ( int i = 0; i < charBlocks.size(); i++ ) {
				blocks[i+taxa.length+treeBlocks.size()] = charBlocks.get(i);
			}
			theBlocks = blocks;
		} catch ( Exception e) {
			e.printStackTrace();
		}
		return theBlocks;
	}
	
	private static void resolveTreeTaxa(HashMap taxaBlocks,Vector treeBlocks){
		for ( int i = 0; i < treeBlocks.size(); i++ ) {
			String otusIdRef = (String)((Forest)treeBlocks.get(i)).getGeneric("otus");
			Taxa taxa = (Taxa)taxaBlocks.get(otusIdRef);
			((Forest)treeBlocks.get(i)).setTaxa(taxa);
			HashMap taxon = new HashMap();
			Containable[] taxonObj = taxa.getEntities();
			for ( int j = 0; j < taxonObj.length; j++ ) {
				taxon.put(taxonObj[j].getXmlId(), taxonObj[j]);
			}
			Containable[] trees = ((Forest)treeBlocks.get(i)).getEntities();
			for ( int j = 0; j < trees.length; j++ ) {
				Containable[] nodes = ((Listable)trees[j]).getEntities();
				for ( int k = 0; k < nodes.length; k++ ) {
					String otuIdRef = (String)nodes[k].getGeneric("otu");
					if ( otuIdRef != null ) {
						Taxon theTaxon = (Taxon)taxon.get(otuIdRef);
						if ( theTaxon != null ) {
							((TaxonLinker)nodes[k]).setTaxon(theTaxon);
						}
					}
				}
			}
		}
	}
	private static void resolveMatrixTaxa(HashMap taxaBlocks,Vector charBlocks){
		for ( int i = 0; i < charBlocks.size(); i++ ) {
			String otusIdRef = (String)((Matrix)charBlocks.get(i)).getGeneric("otus");
			Taxa taxa = (Taxa)taxaBlocks.get(otusIdRef);
			((Matrix)charBlocks.get(i)).setTaxa(taxa);
			HashMap taxon = new HashMap();
			Containable[] taxonObj = taxa.getEntities();
			for ( int j = 0; j < taxonObj.length; j++ ) {
				taxon.put(taxonObj[j].getXmlId(), taxonObj[j]);
			}
			Containable[] rows = ((Matrix)charBlocks.get(i)).getEntities();
			for ( int j = 0; j < rows.length; j++ ) {
				String otuIdRef = (String)rows[j].getGeneric("otu");
				if ( otuIdRef != null ) {
					Taxon theTaxon = (Taxon)taxon.get(otuIdRef);
					((TaxonLinker)rows[j]).setTaxon(theTaxon);
				}
			}
		}		
	}
	
	private static Vector processCharacters(NodeList charElts) throws ObjectMismatch {
		Vector charBlocks = new Vector();
		if ( charElts.getLength() == 0) return charBlocks;
		for ( int i = 0; i < charElts.getLength(); i++ ) {
			Matrix charObj = (Matrix)objFromElement((Element)charElts.item(i));
			String type = ((Element)charElts.item(i)).getAttribute("xsi:type");
			int start = type.indexOf(":");
			int finish = type.lastIndexOf("Seqs");
			if ( finish == -1 ) {
				finish = type.lastIndexOf("Cells");
			}
			type = type.substring(start + 1, finish);
			logger.info(type);
			Datatype to = Datatype.getInstance(type);
			charObj.setTypeObject(to);
			String otus = ((Element)charElts.item(i)).getAttribute("otus");
			charObj.setGeneric("otus", otus);
			HashMap stateSet = new HashMap();
			HashMap stateSetForChar = new HashMap();
			HashMap colIndices = new HashMap();
			HashMap symbolOfId = new HashMap();
			if ( ((Element)charElts.item(i)).getElementsByTagName("format").getLength() > 0 ) {
				Element formatElt = (Element)((Element)charElts.item(i)).getElementsByTagName("format").item(0);
				NodeList statesElts = formatElt.getElementsByTagName("states");
				for ( int j = 0; j < statesElts.getLength(); j++ ) {
					String id = ((Element)statesElts.item(j)).getAttribute("id");
					HashMap lookup = new HashMap();					
					HashMap innerSet = new HashMap();
					resolveMapping(((Element)statesElts.item(j)).getElementsByTagName("state"),symbolOfId,lookup);
					resolveMapping(((Element)statesElts.item(j)).getElementsByTagName("polymorphic_state_set"),symbolOfId,lookup);
					resolveMapping(((Element)statesElts.item(j)).getElementsByTagName("uncertain_state_set"),symbolOfId,lookup);
					innerSet.put("lookup", lookup);
					innerSet.put("symbols", symbolOfId);
					stateSet.put(id, innerSet);
					charObj.setLookup(lookup);
				}
				if ( ((Element)charElts.item(i)).getElementsByTagName("char").getLength() > 0 ) {
					NodeList colElts = ((Element)charElts.item(i)).getElementsByTagName("char");
					for ( int j = 0; j < colElts.getLength(); j++ ) {
						String id = ((Element)colElts.item(j)).getAttribute("id");
						String states = ((Element)colElts.item(j)).getAttribute("states");
						stateSetForChar.put(id, states);
						colIndices.put(id, new Integer(j));
					}
				}				
			}
			Element matrixElt = (Element)((Element)charElts.item(i)).getElementsByTagName("matrix").item(0);
			processMatrix(matrixElt,symbolOfId,colIndices,charObj);
			charBlocks.add(charObj);
		}
		return charBlocks;
	}
	
	private static void processMatrix(Element matrixElt,HashMap symbolOfId,HashMap colIndices,Matrix charObj) throws ObjectMismatch {
		NodeList row = matrixElt.getElementsByTagName("row");
		for ( int i = 0; i < row.getLength(); i++ ) {
			Datum datum = (Datum)objFromElement((Element)row.item(i));
			String otuId = ((Element)row.item(i)).getAttribute("otu");
			datum.setGeneric("otu", otuId);
			datum.setTypeObject(charObj.getTypeObject());
			String[] chars = new String[colIndices.size()];
			charObj.insert(datum);
			NodeList cell = ((Element)row.item(i)).getElementsByTagName("cell");
			if ( cell.getLength() != 0 ) {
				for ( int j = 0; j < cell.getLength(); j++ ) {
					String charId = ((Element)cell.item(j)).getAttribute("char");
					String stateId = ((Element)cell.item(j)).getAttribute("state");
					int charIndex;
					String state;
					if ( colIndices.containsKey(charId) ) {
						charIndex = ((Integer)colIndices.get(charId)).intValue();
					}
					else {
						charIndex = Integer.parseInt(charId);
					}
					if ( symbolOfId.containsKey(stateId) ) {
						state = (String)symbolOfId.get(stateId);
					}
					else {
						state = stateId;
					}
					chars[charIndex] = state;
				}
				String missing = ""+charObj.getMissing();
				for ( int j = 0; j < chars.length; j++ ) {
					if ( chars[j] == null ) {
						chars[j] = missing;
					}
				}
				datum.insert(chars);
			}
			else {
				Element seq = (Element)((Element)row.item(i)).getElementsByTagName("seq").item(0);
				datum.insert(seq.getTextContent());			
			}			
		}		
	}
	
	private static void resolveMapping(NodeList elts,HashMap symbolOfId,HashMap lookup) {
		for ( int i = 0; i < elts.getLength(); i++ ) {
			String symbol = ((Element)elts.item(i)).getAttribute("symbol");
			String id = ((Element)elts.item(i)).getAttribute("id");
			symbolOfId.put(id, symbol);
			NodeList mappings = ((Element)elts.item(i)).getElementsByTagName("mapping");
			Vector symlist = new Vector();
			if ( mappings.getLength() > 0 ) {
				for ( int j = 0; j < mappings.getLength(); j++ ) {
					String refId = ((Element)mappings.item(j)).getAttribute("state");
					String refSym = (String)symbolOfId.get(refId);
					symlist.add(refSym);
				}
			}
			else {
				symlist.add(symbol);				
			}
			lookup.put(symbol, symlist);
		}
	}
	
	private static Tree processTree(Element treeElt) throws ObjectMismatch {
		Tree treeObj = (Tree)objFromElement(treeElt);
		HashMap sourceOf = new HashMap();
		HashMap targetOf = new HashMap();
		HashMap lengthOf = new HashMap();
		HashMap otuOf = new HashMap();
		HashMap nodeById = new HashMap();
		NodeList nodeElts = treeElt.getElementsByTagName("node");
		for ( int i = 0; i < nodeElts.getLength(); i++ ) {
			org.biophylo.Forest.Node nodeObj = (org.biophylo.Forest.Node)objFromElement((Element)nodeElts.item(i));
			String id = nodeObj.getXmlId();
			treeObj.insert(nodeObj);
			nodeById.put(id, nodeObj);
			String otuId = ((Element)nodeElts.item(i)).getAttribute("otu");
			if ( otuId != null ) {
				nodeObj.setGeneric("otu",otuId);
			}
		}
		NodeList edgeElts = treeElt.getElementsByTagName("edge");
		for ( int i = 0; i < edgeElts.getLength(); i++ ) {
			String target = ((Element)edgeElts.item(i)).getAttribute("target");
			String source = ((Element)edgeElts.item(i)).getAttribute("source");
			String length = ((Element)edgeElts.item(i)).getAttribute("length");
			sourceOf.put(target, source);
			targetOf.put(source, target);
			if ( length != null ) {
				lengthOf.put(target, length);
			}
		}
		Containable[] entities = treeObj.getEntities();
		for ( int i = 0; i < entities.length; i++ ) {
			String id = entities[i].getXmlId();
			if ( sourceOf.containsKey(id) ) {
				org.biophylo.Forest.Node parentNode = (org.biophylo.Forest.Node)nodeById.get((String)sourceOf.get(id));
				((org.biophylo.Forest.Node)entities[i]).setParent(parentNode);
				parentNode.setChild((org.biophylo.Forest.Node)entities[i]);
			}
			if ( lengthOf.containsKey(id) ) {
				((org.biophylo.Forest.Node)entities[i]).setBranchLength(Double.parseDouble((String)lengthOf.get(id)));
			}
			if ( otuOf.containsKey(id) ) {
				((org.biophylo.Forest.Node)entities[i]).setGeneric("otu",(String)otuOf.get(id));
			}
		}
		if ( treeElt.getElementsByTagName("rootedge").getLength() > 0 ) {
			Element rootEdge = (Element)treeElt.getElementsByTagName("rootedge").item(0);
			String target = rootEdge.getAttribute("target");
			String bl = rootEdge.getAttribute("length");
			((org.biophylo.Forest.Node)nodeById.get(target)).setBranchLength(Double.parseDouble(bl));
		}
		return treeObj;
	}
	
	private static Vector processTrees(NodeList treesElts) throws ObjectMismatch {
		Vector result = new Vector();
		if(treesElts==null) return result;
		for ( int i = 0; i < treesElts.getLength(); i++ ) {
			Forest forestObj = (Forest)objFromElement((Element)treesElts.item(i));
			String otusId = ((Element)treesElts.item(i)).getAttribute("otus");
			forestObj.setGeneric("otus", otusId);
			NodeList treeElts = ((Element)treesElts.item(i)).getElementsByTagName("tree");
			for ( int j = 0; j < treeElts.getLength(); j++ ) {
				Tree tree = processTree((Element)treeElts.item(j));
				forestObj.insert(tree);
			}
			result.add(forestObj);
		}		
		return result;
	}
	
	private static HashMap processOtus(NodeList otusElts) {
		HashMap taxaBlocks = new HashMap();
		for ( int i = 0; i < otusElts.getLength(); i++ ) {
			Taxa taxa = (Taxa)objFromElement((Element)otusElts.item(i));
			NodeList otuElts = ((Element)otusElts.item(i)).getElementsByTagName("otu");
			for ( int j = 0; j < otuElts.getLength(); j ++ ) {
				Taxon taxon = (Taxon)objFromElement((Element)otuElts.item(j));
				try { 
					taxa.insert(taxon);
				} catch ( Exception e ) {
					e.printStackTrace();
				}
			}
			taxaBlocks.put(taxa.getXmlId(), taxa);
		}
		return taxaBlocks;
	}
	
	private static XMLWritable objFromElement(Element elt) {
		String tagName = elt.getTagName();
		XMLWritable obj = null;
		try {
			obj = (XMLWritable)((Class)factory.get(tagName)).newInstance();
		} catch ( Exception e ) {
			e.printStackTrace();
		}
		obj.setXmlId(elt.getAttribute("id"));
		obj.setName(elt.getAttribute("label"));
		NodeList children = elt.getChildNodes();
		for ( int i = 0; i < children.getLength(); i++ ) {
			if ( children.item(i).getNodeType() == 1 && children.item(i).getNodeName().equals("dict") ) {
				HashMap dict = parseDict((Element)children.item(i));
				obj.setGeneric("dict", dict);
			}
		}
		return obj;
	}
	
	private static HashMap parseDict(Element elt) {
		HashMap result = new HashMap();
		NodeList children = elt.getChildNodes();
		for ( int i = 0; i < children.getLength(); i++ ) {
			if ( children.item(i).getNodeType() == 1 && children.item(i).getNodeName().equals("key") ) {
				String key = children.item(i).getTextContent();
				Element valueElt = null;
				VALUE: for ( int j = i+1; j < children.getLength(); j++ ) {
					if ( children.item(j).getNodeType() == 1 ) {
						valueElt = (Element)children.item(j);
						break VALUE;
					}
				}
				if ( valueElt.getTagName().equals("dict") ) {
					result.put(key, parseDict(valueElt));
				}
				else if ( valueElt.getTagName().indexOf("vector") > 0 ) {
					Vector value = new Vector();
					value.add(valueElt.getTagName());
					String[] valueArray = valueElt.getTextContent().split(" ");
					for ( int k = 0; k < valueArray.length; k++ ) {
						value.add(valueArray[i]);
					}
					result.put(key, value);
				}
				else if ( valueElt.getTagName().equals("any") ) {
					Vector value = new Vector();
					value.add("any");
					value.add(valueElt);
					result.put(key, value);
				}
				else {
					Vector value = new Vector();
					value.add(valueElt.getTagName());
					value.add(valueElt.getTextContent());
					result.put(key, value);
				}
			}
		}		
		return result;
	}

}
